File: 515.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int max(int i, int j) {
        return i > j ? i : j;
    }

    public List<Integer> largestValues(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        perform(root, 0, res);
        return res;
    }

    private void perform(TreeNode r, int level, List<Integer> res) {
        if (r == null)
            return;
        if (res.size() == level)
            res.add(r.val);
        else
            res.set(level, max(res.get(level), r.val));
        perform(r.left, level + 1, res);
        perform(r.right, level + 1, res);
    }
}

File: 515.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def perform(self, r: Optional[TreeNode], level: int, res: List[int]):
        if r is None:
            return
        if len(res) == level:
            res.append(r.val)
        else:
            res[level] = max(res[level], r.val)
        self.perform(r.left, level + 1, res)
        self.perform(r.right, level + 1, res)

    def largestValues(self, r: Optional[TreeNode]) -> List[int]:
        res = []
        self.perform(r, 0, res)
        return res
